parsl/dataflow/dflow.py

```diff
@@ -352,21 +355,36 @@ def handle_exec_update(self, task_record, future):
                     # record the inner app ID in monitoring, and add a completion
                     # listener to that inner future.
 
-                    inner_future = future.result()
+                    joinable = future.result()
 
                     # Fail with a TypeError if the joinapp python body returned
                     # something we can't join on.
-                    if isinstance(inner_future, Future):
+                    if isinstance(joinable, Future):
+                        task_record['status'] = States.joining
+                        task_record['joins'] = joinable
+                        task_record['join_lock'] = threading.Lock()
+                        joinable.add_done_callback(partial(self.handle_join_update, task_record))
+                    elif isinstance(joinable, list):  # TODO: should this be list or arbitrary iterable?
                         task_record['status'] = States.joining
-                        task_record['joins'] = inner_future
-                        inner_future.add_done_callback(partial(self.handle_join_update, task_record))
+                        task_record['joins'] = joinable
+                        task_record['join_lock'] = threading.Lock()
```

Comment:
To avoid a race-condition between adding the callbacks and the processing in handle_join_update, shouldn't we have this set the callback in a locked context?

Reply:
I think it's fine to add callbacks outside of a locked context: the callbacks are safe to fire instantly (when one or more joining applications have already completed) because nothing else happens after this except adding more callbacks.

There is no need to wait for all callbacks to be added before letting the first callback fire and handle things.